### ISSUE TITLE
Test vectors results may be incorrect for Single Step KDF HMAC_hash with default salt 

### ISSUE THREAD

**Issue by slontis** [patrickfav/singlestep-kdf#1]
It looks like you are using the output size as the default salt size instead of the input block size
i.e sha1/sha224/sha256 = 64/64/64 bytes instead of 20/28/32 bytes.

In https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Cr1.pdf Section 4.1

If H(x) = HMAC-hash(salt, x), then, in the absence of an agreed-upon alternative, the 
default_salt shall be an all-zero byte string whose bit length equals that specified as 
the bit length of an **input block** for the hash function.
I am looking at implementing kdf single step for openssl and was looking for some test vectors. I got a different result, and are looking to confirm that I am doing the right thing :)

**Comment by slontis** [comment]
Luckily I have local copies of some of the NIST docs. That link wont work while there is a U.S government funding issue.

**Comment by patrickfav** [comment]
Hi slontis,

Sorry for getting back to you so late, forgot to add myself as "watcher". I will look into this issue and will get back to you.

**Comment by slontis** [comment]
Cheers :)

**Comment by patrickfav** [comment]
You are correct, I use the output length, not the input block length:

https://github.com/patrickfav/singlestep-kdf/blob/9868b76c70adddbb3ccd8ead20f71a4f23556470/src/main/java/at/favre/lib/crypto/HFunction.java#L169

I think the misunderstanding comes from the fact that HKDF, which I implemented previously, uses the out length for omitted salt https://tools.ietf.org/html/rfc5869 (just fyi :))

I will fix the issue and recreate the test vectors in the next days, I'd appreciate feedback from you if the output matches your implementation :)

**Comment by patrickfav** [comment]
Well, changing the null-salt to a byte array of length block size, does not change anything. Looking at the way HMAC handles keys != block size (from [wikipedia](https://en.wikipedia.org/wiki/HMAC)):

```
//Keys longer than blockSize are shortened by hashing them
if (length(key) > blockSize) then
    key ← hash(key) //Key becomes outputSize bytes long
   
//Keys shorter than blockSize are padded to blockSize by padding with zeros on the right
if (length(key) < blockSize) then
    key ← Pad(key, blockSize) //pad key with zeros to make it blockSize bytes long
```

As you can see it doesn't matter if I pass a zero filled byte array with length 32 or 64, it will always end up with a 64 byte long zero filled byte array (when used with SHA-256 e.g.). @slontis do you maybe have another idea why our results differ?

**Comment by patrickfav** [comment]
Referenced in c50a9b7cf1017d960f7148de0f93482083c85279

**Comment by patrickfav** [comment]
Hey @slontis, any idea why our test results differ?

**Comment by slontis** [comment]
I will have a look..

**Comment by slontis** [comment]
You appear to be correct about HMAC.
I have refactored my code since I logged this issue (which was quite a while ago). I remember changing my code so that it gave the same results as your test vectors.
It looks like I get the same result - so there must of been something wrong in my initial implementation. (If I spot anything I will let you know).
So assume your test vectors are correct. Sorry for the confusion.
